tsl/src/nodes/decompress_chunk/decompress_chunk.c

```diff
@@ -359,6 +374,184 @@ cost_decompress_chunk(Path *path, Path *compressed_path)
 	path->rows = compressed_path->rows * DECOMPRESS_CHUNK_BATCH_SIZE;
 }
 
+/*
+ * Calculate the costs for retrieving the decompressed in-order using
+ * a binary heap.
+ */
+static void
+cost_decompress_chunk_merge(PlannerInfo *root, DecompressChunkPath *dcpath, Path *child_path)
+{
+	Path sort_path; /* dummy for result of cost_sort */
+
+	cost_sort(&sort_path,
+			  root,
+			  dcpath->compressed_pathkeys,
+			  child_path->total_cost,
+			  child_path->rows,
+			  child_path->pathtarget->width,
+			  0.0,
+			  work_mem,
+			  -1);
+
+	/* startup_cost is cost before fetching first tuple */
+	//if (sort_path.rows > 0)
+	//	dcpath->cpath.path.startup_cost = sort_path.total_cost / sort_path.rows;
+
+	/*
+	The cost model for the normal chunk decompression produces the following total
+	costs.
+
+	Segments  Total costs
+	 10         711.84
+	 50        4060.91
+	 100       8588.32
+	 10000   119281.84
+
+	The cost model of the regular decompression is roughly linear. Opening multiple batches in
+	parallel needs resources and merging a high amount of batches becomes inefficient at some point.
+	So, we use a quadratic cost model here to have higher costs than the normal decompression when
+	more than ~100 batches are used. We set DECOMPRESS_CHUNK_HEAP_MERGE_CPU_TUPLE_COST to 0.8
+	to become most costly as soon as we have to process more than 120 batches.
+
+	 Note: Don't take the number of tuples into consideration here because of issue #5481
+	 */
+	dcpath->cpath.path.total_cost =
+		sort_path.total_cost + pow(sort_path.rows, 2) * DECOMPRESS_CHUNK_HEAP_MERGE_CPU_TUPLE_COST;
```

Comment:
`R` is the number of rows.

take a look at:
```
select attname,n_distinct from pg_stats where tablename='stage1';
```

I think we can access that info (somehow :) )

Reply:
I did some experiments with the `n_distinct` value. In my experiments, it was often populated with `-1` (this indicates that the number of distinct values is the same as the number of rows). So, it didn't give a more precise bound for the number of parallel open batches. 

I suggest changing this later if we notice that the optimization is not picked up much in production.

Reply:
yes; but I guess you either distinct values in the `segmentby` column; or you didn't had stats on the compressed table

it should work; and for a random table at my system it does work:

```
tsdb=# select attname,n_distinct from pg_stats where tablename='compress_hyper_187_3980_chunk' and attname='bssid';
 attname | n_distinct 
---------+------------
 bssid   |          3
(1 row)

tsdb=# select bssid,count(1) from _timescaledb_internal.compress_hyper_187_3980_chunk group by bssid;
       bssid       | count 
-------------------+-------
 A0:B1:C5:D2:E0:F3 |    21
 01:02:03:04:05:06 |     6
 22:32:A2:B3:05:98 |    21
(3 rows)
```

Reply:
I think instead of using `n**2` cost ; you should use the right cost function

I think the cost function for this operation should be:
`output_row_count * log(input_row_count) * some_const`


and on the other hand not expose the path to the planner in cases when you think it would not work.

If you can estimate the memory needed to have 1 compressed row in-flight processing => you should divide the workmem you think this operator may take up (configure knob?)  to get the max number of input rows you think its ok to take - so that you can bail out in cases you don't want to handle

Reply:
I think we can optimize the costing in a follow-up PR and make adjustments based on user feedback. I agree that  we should also factor in work_mem when deciding about this optimization.